curator-client/src/main/java/org/apache/curator/ConnectionState.java

```diff
@@ -160,13 +162,33 @@ public void process(WatchedEvent event)
             }
         }
 
+        // only wait during tests
+        assert waitOnExpiredEvent(event.getState());
```

Comment:
I changed using a debug flag instead of assert. I added an accessor class to be able to set the flag in the package private ConnectionState class.

I cannot cause the problem to occur without this code.